### ISSUE TITLE
Foreign keys

### ISSUE THREAD

**Issue by FlipEverything** [spotorm/spot2#77]
Hello!

I have to say Spot is a very neat library. I set up my database schema with it and everything was fine until I realized that the foreign keys are missing from the db. Am I blind or something? If I create the foreign keys by hand the next migration removes them. Do Spot support foreign keys? If not then why not? 

What should I do? Switch to a different orm (datamapper) or extend Spot's functionality?

PS.: I need foreign keys for my university project.

**Comment by tuupola** [comment]
Do not create them by hand. Create them with the migrations. For example.

```php
public static function fields()
{
    return [
        "id"        => ["type" => "integer", "primary" => true, "autoincrement" => true],
        "parent_id" => ["type" => "integer"],
        ...
        "created_at" => ["type" => "datetime", "value" => new \DateTime()],
        "updated_at" => ["type" => "datetime", "value" => new \DateTime()]
    ];
}
```

**Comment by FlipEverything** [comment]
That's okay. I have an Entity like this:

```php
public static function fields()
  {
    return [
      'id' => ['type' => 'integer', 'primary' => true, 'autoincrement' => true],
      'course_id' => ['type' => 'integer', 'index' => true],
      'student_id' => ['type' => 'integer', 'index' => true]
    ];
  }
  public static function relations(Mapper $mapper, Entity $entity)
  {
    return [
        'course' => $mapper->belongsTo($entity, 'Entity\Course', 'course_id'),
        'student' => $mapper->belongsTo($entity, 'Entity\Student', 'student_id')
    ];
  }
```


The field part is okay, Spot created the table from it. 
But the thing is the 'relations' part is not mapped to the database. I need constraints like this:

```sql
CREATE TABLE example (
    id INT, 
    course_id INT,
    student_id INT,
    INDEX course_id_fk (course_id),
    INDEX student_id_fk (student_id),
    FOREIGN KEY (course_id) 
        REFERENCES course(id)
        ON DELETE CASCADE,
    FOREIGN KEY (student_id) 
        REFERENCES student(id)
        ON DELETE CASCADE
) ENGINE=INNODB;
```
But Spot only did this:
```sql
CREATE TABLE example (
    id INT, 
    course_id INT,
    student_id INT,
    INDEX course_id (course_id),
    INDEX student_id (student_id)
) ENGINE=INNODB;
```

**Comment by vlucas** [comment]
Foreign keys are something that Spot simply doesn't have yet - not because of any specific decision that was made to exclude them. It would be great if you could fork and modify Spot to include them by default (and optionally omit them in case of errors).

The file you will want to look at to achieve this is here:
https://github.com/vlucas/spot2/blob/master/lib/Spot/Query/Resolver.php#L78

**Comment by nebulousGirl** [comment]
I need the foreign keys too:

Here is a draft of a solution for belongsTo relations, I might do a fork when I'm satisfied with my solution. Insert the code into the <code>migrateCreateSchema()</code> method.

``` php
        
        //TODO Should move that to the entityManager
        $relations = call_user_func(array($entityName, 'relations'), $this->mapper, new $entityName);
        foreach ($relations as $relation_name => $relation) {
            if ($relation instanceof \Spot\Relation\BelongsTo) {
                $parent_entity = $relation->entityName();
                
                $foreignTable = call_user_func(array($parent_entity, 'table'));
                //TODO Should be customisable, sometimes you want to delete other times you don't want to
                $table->addForeignKeyConstraint($foreignTable, array($relation->localKey()), array($relation->foreignKey()), array("onDelete" => "CASCADE"));
                
            }
            //TODO Other types of relations
        }
```

**Comment by vlucas** [comment]
Thanks @nebulousGirl - that's a great start.

**Comment by Artamiel** [comment]
Bumping here.

I've decided to put some work on this issue, because having foreign keys supported would be really a good thing. Following the example from above everything should be working just fine when it comes to `BelongsTo` and `HasOne` relations and I've added some code to my fork of this project with the ability to define yourself what should happen when `onUpdate` and `onDelete` occurs. But, when I ran the tests, I noticed a strange behaviour for `HasOne` relation.

The relation itself is added at the inverse side of the relation, and it should be at the owning side. Because when we execute `addForeignKeyConstraint` the code is expecting for the column with our foreign key to be defined in this table, which would eventually result to `Invalid Foreign key Consrtraint` exception. At the very least I can say that currently, if this feature is enabled, [this test](https://github.com/vlucas/spot2/blob/master/tests/Entity/Event.php#L47) will fail.

You can see for a reference, the way **Doctrine** handles this for better understanding of what I'm trying to say. Anyway, any thoughts on this @vlucas ?

**Comment by FlipEverything** [comment]
Do we have something here? Is @nebulousGirl's solution any good or do I need to look into it?

I invested a lot of time in Spot during the past few months. I built my own library around it, but now I have a lot of problems caused by the lack of foreign keys. I can't migrate my tables properly and now it's a really big pile of painful mess. I don't like the idea of dropping the database constraints.

**Comment by FlipEverything** [comment]
I tested @nebulousGirl code and fixed it: 3057994

1. The relations variable was always an empty array for me, it didn't do anything.
2. Added alias support.

It works for me but further testing is required.
I'm using my fork now but If you want I'll create a PR.

**Comment by tuupola** [comment]
#77 and #156 have been merged.
